src/AutoMapper/QueryableExtensions/ExpressionBuilder.cs

```diff
@@ -37,6 +37,7 @@ public class ExpressionBuilder : IExpressionBuilder
             new NullableDestinationExpressionBinder(),
             new NullableSourceExpressionBinder(),
             new AssignableExpressionBinder(),
+            new ImplicitAssignableExpressionBinder(),
```

Comment:
Can this be done elsewhere? Because EF6 doesn't allow just any cast, so this could break things.

Reply:
Hard to decide, 2 Test cases fail if I remove it (Project_to_custom_struct and Project_to_custom_struct_with_mapping). But, if entity framework is used, this case should not occour, cause entity framework also doesn't support custom struct types, right?

Reply:
I don't know about that, but the main use case for ProjectTo is EF, so it seems to me like this code should not be here.

Reply:
@TylerCarlson1 @BlaiseD Wouldn't it be possible to achieve the same effect and have this code elsewhere, not on the EF path?

Reply:
I think you would need to use a different ExpressionVisitor with new extension methods to avoid affecting other users.  It would also help to create an issue to clarify the goal here.

Reply:
I think that extension method can be UseAsDataSource.

Reply:
@lbargaoanu My best guess right now would be make the read only array be a list like IMapper list is so you can inject the ImplicitAssignableExpressionBinder at the configuration stage.
Like you can remove assignable mapper if you want if you want explicit cloning, or adding the Collection.Mappers stuff.

Then from there can go about making extension methods saying support OData projection or something along those lines and later on can make one for EF.Core perhaps.

Reply:
Sounds good to me :) @jbogard Thoughts?

Reply:
@BlaiseD Sorry that i didn't open multiple issues to this case. Everything depends on supporting the mapping of custom struct types and the implicit cast to this types if operator-methods are defined. When I find a not supported case, I create a unit-test and try to fix it as best as I can.

The ImplicitAssignableExpressionBinder checks in addition to the other binders, if a type can be convertet implicit.

Reply:
FYI - The following tests fails against ReflectionHelper.CanImplicitConvert.  There's a suggested fix below -  may be incomplete also.

```csharp
    public class CanImplicitConvertFailedTests
    {
        [Fact]
        public void IListOfString_is_assignable_from_ListOfString()
        {
            IList<string> list = new List<string>();
            //Assert
            Assert.True(typeof(List<string>).CanImplicitConvert(typeof(IList<string>)));
        }

        [Fact]
        public void Int_is_assignable_from_short()
        {
            int i = (short)0;
            //Assert
            Assert.True(typeof(short).CanImplicitConvert(typeof(int)));
        }

        [Fact]
        public void Nullable_custom_string_is_assignable_from_custom_string()
        {
            CustString? ss = new CustString("");
            //Assert
            Assert.True(typeof(CustString).CanImplicitConvert(typeof(CustString?)));
        }

        [Fact]
        public void Nullable_int_digit_is_assignable_from_short()
        {
            IntDigit? i = (short)0;
            //Assert
            Assert.True(typeof(short).CanImplicitConvert(typeof(IntDigit?)));
        }

        [Fact]
        public void Nullable_int_is_assignable_from_int()
        {
            int? i = int.MaxValue;
            //Assert
            Assert.True(typeof(int).CanImplicitConvert(typeof(int?)));
        }

        [Fact]
        public void Nullable_int_is_assignable_from_nullable_short()
        {
            int? i = new short?();
            //Assert
            Assert.True(typeof(short?).CanImplicitConvert(typeof(int?)));
        }

        [Fact]
        public void Nullable_int_is_assignable_from_short()
        {
            int? i = (short)0;
            //Assert
            Assert.True(typeof(short).CanImplicitConvert(typeof(int?)));
        }

        [Fact]
        public void Short_is_assignable_from_sbyte()
        {
            short s = (sbyte)0;
            //Assert
            Assert.True(typeof(sbyte).CanImplicitConvert(typeof(short)));
        }
    }

    public struct CustString
    {
        public CustString(string i) { val = i; }
        public string val;

        public static implicit operator string(CustString i)
        {
            return i.val;
        }

        public static implicit operator CustString(string i)
        {
            return new CustString(i);
        }
    }

    public struct IntDigit
    {
        public IntDigit(int i) { val = i; }
        public int val;

        public static implicit operator int(IntDigit i)
        {
            return i.val;
        }

        public static implicit operator IntDigit(int i)
        {
            return new IntDigit(i);
        }
    }

    internal static class Helpers
    {
        //suggested extension
        public static bool CanImplicitConvert(this Type from, Type to)
        {
            if (to.IsAssignableFrom(from))
                return true;

            if (!(!to.IsNullable() && from.IsNullable()))
            {//Anything but To is NOT nullable and From IS nullable
                to = to.IsNullable() ? Nullable.GetUnderlyingType(to) : to;
                from = from.IsNullable() ? Nullable.GetUnderlyingType(from) : from;

                if (dict.ContainsKey(to) && dict[to].Contains(from))
                    return true;
            }

            bool ReturnTypeValid(Type returnType) => returnType == to || (dict.ContainsKey(to) && dict[to].Contains(returnType));
            bool ParameterValid(Type parameterType) => (parameterType == from) || (dict.ContainsKey(parameterType) && dict[parameterType].Contains(from));

            return from.GetMethods(BindingFlags.Public | BindingFlags.Static)
                            .Any(m => m.Name == "op_Implicit"
                                    && ReturnTypeValid(m.ReturnType)
                                    && ParameterValid(m.GetParameters().Single().ParameterType))
                            ||
                            to.GetMethods(BindingFlags.Public | BindingFlags.Static)
                            .Any(m => m.Name == "op_Implicit"
                                    && ReturnTypeValid(m.ReturnType)
                                    && ParameterValid(m.GetParameters().Single().ParameterType));
        }

        private static bool IsNullable(this Type type)
        {
            return type.IsGenericType && type.GetGenericTypeDefinition().Equals(typeof(Nullable<>));
        }

        static Dictionary<Type, HashSet<Type>> dict = new Dictionary<Type, HashSet<Type>>()
        {
            { typeof(decimal), new HashSet<Type> { typeof(byte), typeof(sbyte), typeof(char), typeof(short), typeof(ushort), typeof(int), typeof(uint), typeof(long), typeof(ulong) } },
            { typeof(double), new HashSet<Type> { typeof(byte), typeof(sbyte), typeof(char), typeof(short), typeof(ushort), typeof(int), typeof(uint), typeof(long), typeof(ulong), typeof(float) } },
            { typeof(float), new HashSet<Type> { typeof(byte), typeof(sbyte), typeof(char), typeof(short), typeof(ushort), typeof(int), typeof(uint), typeof(long), typeof(ulong) } },
            { typeof(ulong), new HashSet<Type> { typeof(byte), typeof(char), typeof(ushort), typeof(uint) } },
            { typeof(long), new HashSet<Type> { typeof(byte), typeof(sbyte), typeof(char), typeof(short), typeof(ushort), typeof(int), typeof(uint) } },
            { typeof(uint), new HashSet<Type> { typeof(byte), typeof(char), typeof(ushort) } },
            { typeof(int), new HashSet<Type> { typeof(byte), typeof(sbyte), typeof(char), typeof(short), typeof(ushort) } },
            { typeof(ushort), new HashSet<Type> { typeof(byte), typeof(char) } },
            { typeof(short), new HashSet<Type> { typeof(byte), typeof(sbyte) } }
        };
    }
```

Reply:
Hello,
I added your tests and got an error in AutoMapper.UnitTests.Projection.NestedAndArraysTests.LinqTests.Example

> Testname:	AutoMapper.UnitTests.Projection.NestedAndArraysTests.LinqTests.Example
> Test FullName:	AutoMapper.UnitTests.Projection.NestedAndArraysTests.LinqTests.Example
> Testquelle:	C:\Git-Repositories\AutoMapper\src\UnitTests\Projection\NestedAndArraysTests.cs > Zeile 49
> Testergebnis:	Fehlgeschlagene
>Testdauer:	0:00:00,008

> Ergebnis StackTrace:	
 bei lambda_method(Closure , Entity )
   bei AutoMapper.UnitTests.Projection.NestedAndArraysTests.LinqTests.Example() in C:\Git-Repositories\AutoMapper\src\UnitTests\Projection\NestedAndArraysTests.cs:Zeile 76.
Ergebnis Meldung:	System.InvalidCastException : Das Objekt des Typs "WhereSelectEnumerableIterator`2[AutoMapper.UnitTests.Projection.NestedAndArraysTests.LinqTests+SubEntity,System.String]" kann nicht in Typ "System.String[]" umgewandelt werden.

The reason is that my ImplicitAssignableExpressionBinder is returned because of the 
```csharp
if (to.IsAssignableFrom(from))
                return true;
```

Imo this check shouldn't be part of the Check, because the assignable binding is done by the AssignableExpressionBinder. When the check is removed, ImplicitAssignableExpressionBinder is not returned for the array-type and the test succeed, but some of your new Nullable-Tests fail. So I decided to fix the error by keeping the IsAssignableFrom-check but to move the ImplizitAssignableExpressionBinder to the end of the Bindings-List. This way the required EnumerableExpressionBinder has higher priority and all tests succeed.